src/fluree/db/json_ld/shacl.cljc

```diff
@@ -48,29 +49,70 @@
 
+(defn throw-property-shape-exception!
+  [msg]
+  (throw (ex-info (str "SHACL PropertyShape exception - " msg ".")
+                      {:status 400 :error :db/shacl-validation})))
+
 (defn validate-property
-  "Validates a PropertyShape against a set of flakes"
+  "Validates a PropertyShape for a single predicate
+  against a set of flakes"
   [{:keys [min-count max-count node-kind]} p-flakes]
   (let [n (count p-flakes)]
     (when (and min-count
                (> min-count n))
-      (throw (ex-info (str "SHACL PropertyShape exception - sh:minCount of " min-count
-                           " higher than actual count of " n ".")
-                      {:status 400 :error :db/shacl-validation})))
+      (throw-property-shape-exception! (str "sh:minCount of " min-count " higher than actual count of " n)))
     (when (and max-count
                (> n max-count))
-      (throw (ex-info (str "SHACL PropertyShape exception - sh:maxCount of " max-count
-                           " lower than actual count of " n ".")
-                      {:status 400 :error :db/shacl-validation})))))
+      (throw-property-shape-exception! (str "sh:maxCount of " max-count " lower than actual count of " n)))))
+
+(defn validate-pair-property
+  "Validates a PropertyShape that compares values
+  for a pair of predicates."
+  [{:keys [pair-constraint] :as _p-shape} lhs-flakes rhs-flakes]
+  (let [flake-value (fn [flake] [(flake/o flake) (flake/dt flake)])]
+    (case pair-constraint
+
+      (:equals :disjoint) (let [lhs-values (into #{} (map flake-value) lhs-flakes)
+                                rhs-values (into #{} (map flake-value) rhs-flakes)]
+                            (case pair-constraint
+                              :equals
+                              (when (not= lhs-values rhs-values)
+                                (throw-property-shape-exception!
+                                 (str "sh:equals: " (mapv flake/o lhs-flakes) " not equal to " (mapv flake/o rhs-flakes))))
+                              :disjoint
+                              (when (seq (set/intersection lhs-values rhs-values))
+                                (throw-property-shape-exception!
+                                 (str "sh:disjoint: " (mapv flake/o lhs-flakes) " not disjoint from " (mapv flake/o rhs-flakes))))))
+
+      (:lessThan :lessThanOrEquals) (let [allowed-cmp-results (cond-> #{-1}
+                                                                (= pair-constraint :lessThanOrEquals) (conj 0))]
+                                      (doseq [l-flake lhs-flakes
+                                              r-flake rhs-flakes
+                                              :let [[l-flake-o l-flake-dt] (flake-value l-flake)
+                                                    [r-flake-o r-flake-dt] (flake-value r-flake)]]
+                                        (when (or (not= l-flake-dt
+                                                        r-flake-dt)
```

Comment:
We _do_ need to inspect the datatypes for compatibility in `sh:lessThan` and `sh:lessThanOrEquals` , to avoid nonsense results ( eg `cmp-obj` it gives a "false positive" value of `-1` if you compare a `xsd:string` to a `xsd:long` for example), but I didn't get too sophisticated with any notions of "compatibility" and checked if they were equal. However, again the actual Right Way to do this feels like an open question to me.

Reply:
On second thought, I do think that datatypes merit a bit more care than this for `sh:lessThan` and `sh:lessThanOrEquals`.

This implementation will allow eg two subject id values, or two strings, as long as `cmp-obj` returns `-1`.

I think we want this to only succeed for numeric types, and dates/times.

Reply:
I constrained which types will get compared in https://github.com/fluree/db/pull/401/commits/ff46055b2c770c8db40c215aeaee898bdb770647. I just went with enumerating the valid type constants, but would love to know if there's a better way

Reply:
I don't know of a better way than what you've done.